pkg/cmd/cli/cmd/rsync/rsync.go

```diff
@@ -229,7 +236,76 @@ func (o *RsyncOptions) Validate() error {
 
 // RunRsync copies files from source to destination
 func (o *RsyncOptions) RunRsync() error {
-	return o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	err := o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	if err != nil || !o.Watch {
+		return err
+	}
+
+	dirty := false
+	lastChange := time.Now()
+	var watchError error
+	changeLock := sync.Mutex{}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				glog.V(5).Infof("filesystem watch event: %s", event)
+				lastChange = time.Now()
+				dirty = true
+				changeLock.Unlock()
+
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					watcher.Remove(event.Name)
+				} else {
+					fsnotification.AddRecursiveWatch(watcher, event.Name)
+				}
+			case err := <-watcher.Errors:
+				changeLock.Lock()
+				watchError = fmt.Errorf("error watching filesystem for changes: %v", err)
+				changeLock.Unlock()
+			}
+		}
+	}()
+
+	err = fsnotification.AddRecursiveWatch(watcher, o.Source.Path)
+	if err != nil {
+		// If this fails, it may be necessary to increase available watches on linux
+		// by running 'sudo sysctl fs.inotify.max_user_watches=65536'
+		return fmt.Errorf("error watching source path %s: %v", o.Source.Path, err)
+	}
+
+	for {
+		changeLock.Lock()
+		if watchError != nil {
+			return watchError
+		}
+		// if a change happened more than 2 seconds ago, sync it now.
+		// if a change happened less than 2 seconds ago, sleep for 2 seconds
+		// and see if more changes happen, we don't want to sync when
+		// the filesystem is in the middle of changing due to a massive
+		// set of changes (such as a local build in progress).
+		if dirty && time.Now().After(lastChange.Add(2*time.Second)) {
+			glog.V(1).Info("Synchronizing filesystem changes...")
+			err = o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+			if err != nil {
+				return err
+			}
+			glog.V(1).Info("Done.")
+			dirty = false
+		}
+		changeLock.Unlock()
+		time.Sleep(2 * time.Second)
+	}
+	// unreachable
+	return nil
```

Comment:
The two lines above can/should be removed. The compiler knows about unreachable code, and so does `go vet`:

```console
$ go vet pkg/cmd/cli/cmd/rsync/rsync.go
pkg/cmd/cli/cmd/rsync/rsync.go:308: unreachable code
exit status 1
```